compose: Add warning banner for search views.

josiah-tesfu opened the pull request:
Added a warning compose banner when composing a message in a search view that doesn't include all the messages in a conversation.

This is helpful in a situation where a user sends a message without the full context and forgets they are in a narrowed search view.

Fixes: #25893.

![issue-25893-rec](https://github.com/zulip/zulip/assets/71205057/d4a01fc9-2dab-44d3-bd3b-2f42af87cbac)

Need to confirm the following:
- The condition for a narrow that doesn't show all messages in a conversation is based on the function is_common_narrow() in filter.js. I thought this might be more appropriate than supports_collapsing_recipients() but just wanted to confirm.
- Similarly, the "Go to conversation" button uses parameters defined in get_url_encoding() in filter.js, which I also think is appropriate but would like to double-check.


<details>
<summary>Self-review checklist</summary>


- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

karlstolley commented:
This is looking good, but make sure you run the linter to clean up some whitespace and newlines you've introduced.

josiah-tesfu commented:
Ready for review!

evykassirer commented on web/src/narrow.js:
```diff
@@ -541,6 +543,14 @@ export function activate(raw_operators, opts) {
 
+        // Clears 'compose_in_search_view' when manually exiting a
+        // search view narrow.
+        compose_banner.clear_in_search_view();
```
Why did you need to do this here instead of where the rest of the warnings are cleared in `clear_warnings`?

josiah-tesfu replied:
I didn't do it this way previously and got a bug where after `clear_warnings` was called, the `WARNING` CSS label was gone and couldn't be restored. For some reason, when I cleared the banner this way, I was still able to create a warning banner in the lines after. That's why I updated clear_warnings to have `.not(".compose_in_search_view").remove();`.

Weird bug. Not 100% sure why it worked that way, but it did.

josiah-tesfu commented:
@evykassirer I addressed the comment you left regarding `clear_warnings`. However, I'm completely lost on how to resolve the node test I'm currently failing. It's failing due to  the same line, `compose_banner.clear_in_search_view();`.

evykassirer commented:
I'll take a look, and also will try to figure out the issue I commented on before.

evykassirer commented:
I've opened a separate draft PR (#26205) with a few changes after working on it locally -- you're welcome to copy from there and ask any questions about it -- if you bring those changes over here I can just close the PR.

* The button in the banner wasn't working in this PR. Try it for a result that's not a stream (e.g. just "box") and you'll see it redirects to the homepage instead. `narrow_state.operators()` doesn't work because that contains the search results instead of the details of the message being composed. I changed it to look for the stream and topic in the compose box.
* I moved the opening of the banner to when the composebox is opened, instead of when the narrow is activated which happens only once per search and was why the message wasn't appearing again after you closed it once.

josiah-tesfu commented:
@evykassirer, thanks for the help! I moved over the changes you made. But it looks like the code is now failing a puppeteer test,
`The Puppeteer frontend tests failed! The failing test was:
    ./tools/test-js-with-puppeteer compose.test.ts`.

evykassirer commented on web/src/compose.js:
```diff
@@ -507,6 +509,21 @@ export function initialize() {
+            event.preventDefault();
+
+            const filter = new Filter(narrow_state.operators());
+            window.location.href = filter.generate_redirect_url();
```
You haven't updated this part yet -- see my previous comment about this not working for plain text queries.

evykassirer commented:
There's an error message in the failure saying:

`Cannot read properties of undefined (reading 'is_common_narrow') (undefined:undefined:undefined)`

josiah-tesfu commented:
Missed that one, thanks! But the puppeteer test still fails regardless.

karlstolley commented:
@josiah-tesfu did you mean to close this PR?

evykassirer commented:
He said he didn't, so not sure what happened. He opened a new one here: #26214
